django/db/models/functions/json.py

```diff
@@ -0,0 +1,196 @@
+from django.db.models.constants import LOOKUP_SEP
+from django.db.models.expressions import Func, Value
+from django.db.models.fields.json import JSONField, compile_json_path
+from django.db.models.functions import Cast
+
+
+class JSONSet(Func):
+    output_field = JSONField()
+
+    def __init__(self, field_name, **fields):
+        if not fields:
+            raise TypeError("JSONSet requires fields parameter")
+        self.fields = fields
+        super().__init__(field_name)
+
+    def as_sql(
+        self,
+        compiler,
+        connection,
+        function=None,
+        template=None,
+        arg_joiner=None,
+        **extra_context,
+    ):
+        copy = self.copy()
+        new_source_expression = copy.get_source_expressions()
+
+        for key, value in self.fields.items():
+            key_paths = key.split(LOOKUP_SEP)
+            key_paths_join = compile_json_path(key_paths)
+            new_source_expression.extend(
+                (
+                    Value(key_paths_join),
+                    Cast(
+                        Value(value, output_field=self.output_field),
+                        output_field=self.output_field,
+                    ),
+                )
+            )
+
+        copy.set_source_expressions(new_source_expression)
+
+        return super(JSONSet, copy).as_sql(
+            compiler,
+            connection,
+            function="JSON_SET",
+            **extra_context,
+        )
+
+    def as_postgresql(self, compiler, connection, **extra_context):
+        copy = self.copy()
+
+        all_items = list(self.fields.items())
+        key, value = all_items[0]
+        rest = all_items[1:]
+
+        if rest:
```

Comment:
This is worth a code comment,e.g.

```suggestion
        # JSONB_SET does not support arbitrary number of arguments,
        # so convert multiple updates into recursive calls.
        if rest:
```